recitation_2/bit_operators.md

```diff
@@ -1,26 +1,428 @@
+{
+    printf("Goodbye, %s!\n", s);
+}
+
+```
+
+Because `hello()` and `goodbye()` are not defined in `myprogram.c`, we need to include their respective header files in `myprogram.c`  to let our program know that they are defined elsewhere:
+
+`hello.h`:
+
+```c
+#ifndef _HELLO_H_
+#define _HELLO_H_
+
+void hello(char *s);
+
+#endif
+
+```
+
+`goodbye.h`:
+
+```c
+#ifndef _GOODBYE_H_
+#define _GOODBYE_H_
+
+void goodbye(char *s);
+
+#endif
+
+```
+
+- Once the header is included, it checks if a unique value (in this case `_GOODBYE_H_`) is defined. Then if it's not defined, it defines it and continues. If the code is included again, the first `ifndef` fails, resulting in a blank file. This is called an include guard and prevents double declarations.
+
+### Compiling
+
+```bash
+$ ls
+goodbye.c       goodbye.h       hello.c         hello.h         myprogram.c
+```
+
+The `.c` files need to be compiled into object files (`.o` files).
+
+We can do this by using the `gcc -c` command.
+
+```bash
+$ gcc -c -o hello.o hello.c 
+$ gcc -c -o goodbye.o goodbye.c 
+$ gcc -c -o myprogram.o myprogram.c 
+```
+
+- `gcc` is the C compiler that we use
+- `-c` specifies that we want to compile
+- `-o <filename>.o` specifies that we want the output (the object file) to be named `<filename>.o`
+  - The `.o` file should have the same name as the `.c` file
+- `<filename>.c` is the `.c` file that we want to compile
+
+Format: `gcc -c -o <output name> <ingredients>`
+
+Now, we have the object files:
+
+```bash
+$ ls
+goodbye.c       goodbye.o       hello.h         myprogram.c
+goodbye.h       hello.c         hello.o         myprogram.o
+```
+
+- `gcc -c` will pre-process (ie. including the headers), compile, and assemble the files; taking in a `.c` file and outputing a `.o` file.
+
+### Linking
+
+Now, we need to link the `.o` files to create our executable.
+
+To link:
+
+```bash
+$ gcc -o myprogram hello.o goodbye.o myprogram.o
+```
+
+Format: `gcc -o <output name> <ingredients`
+
+Now we have our executable, `myprogram`:
+
+```bash
+$ ls
+goodbye.c       goodbye.o       hello.h         myprogram       myprogram.o
+goodbye.h       hello.c         hello.o         myprogram.c
+```
+
+```bash
+$ ./myprogram Faustina
+This is my program!
+Hello, Faustina! 
+Goodbye, Faustina!
+```
+
+### Makefile
+
+This process to produce our executable required many steps. We can streamline it by creating a `Makefile`, which give us a simple way to organize code compilation.
+
+Because `hello.o` is produced by including `hello.h` and compiling `hello.c`, we can say that `hello.o` depends on `hello.c` and `hello.h`. Similarly, `myprogram` is produced by linking `hello.o`, `goodbye.o`, and `myprogram.o`, so we say that `myprogram` depends on these object files.
+
+We should only need to rebuild something if one of the files it depends on has changed. After all, part of the reason for separate compilation is to avoid the cost of rebuilding things we do not need to rebuild.
```

Comment:
```suggestion
We should only need to rebuild something if at least one of the files it depends on has changed. After all, part of the reason for separate compilation is to avoid the cost of rebuilding things we do not need to rebuild.
```